gswrap/__init__.py

```
@@ -0,0 +1,626 @@
+        https://cloud.google.com/storage/docs/gsutil/commands/ls
+
+        :param gcs_url: Google cloud storage URL
+        :param recursive: List only direct subdirectories
+        :return: List of Google cloud storage URLs according the given URL
+        """
+        url = _UniformPath(res_loc=gcs_url)
+
+        if not url.is_cloud_url:
+            raise ValueError(
+                "{} is not a google cloud storage URL".format(gcs_url))
+
+        blobs = self._ls(url=url, recursive=recursive)
+
+        blob_list = []
+        for element in blobs:
+            blob_list.append('gs://' + url.bucket + '/' + element)
+
+        return blob_list
+
+    def _ls(self, url: _UniformPath, recursive: bool = False) -> List[str]:
+        """
+        List the files on Google Cloud Storage given the prefix.
+
+        :param url: uniform google cloud url
+        :param recursive: List only direct subdirectories
+        :return: List of prefixes according to the given prefix
+        """
+        if recursive:
+            delimiter = ''
+        else:
+            delimiter = '/'
+        # add trailing backslash to limit search to this file/folder
+        # Google Cloud Storage prefix have no leading backslash
+        raw_prefix = url.prefix.as_posix(
+            add_trailing_backslash=url.prefix.had_trailing_backslash,
+            remove_leading_backslash=url.is_cloud_url)
+
+        self._change_bucket(bucket_name=url.bucket)
+        is_not_blob = raw_prefix == "" or self._bucket.get_blob(
+            blob_name=raw_prefix) is None
+
+        prefix = url.prefix.as_posix(
+            add_trailing_backslash=is_not_blob,
+            remove_leading_backslash=url.is_cloud_url)
+        iterator = self._bucket.list_blobs(
+            versions=True, prefix=prefix, delimiter=delimiter)
+
+        blob_names = _list_blobs(iterator=iterator)
+
+        if not blob_names:
+            raise google.api_core.exceptions.GoogleAPIError(
```

Do we really want to return an error if nothing was listed? I would actually just return an empty list.